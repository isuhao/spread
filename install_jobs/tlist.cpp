#include "tlist.hpp"

using namespace Spread;

void TargetList::remove(const Hash::DirMap &files)
{
  LockGuard lock(mutex);
  Hash::DirMap::const_iterator it;
  for(it = files.begin(); it != files.end(); it++)
    {
      HMap::iterator it2 = list.find(it->second);
      if(it2 != list.end())
        {
          assert(it2->second);
          list.erase(it2);
        }
    }
}

// Inserts element if missing. Returns true if found, false if
// inserted.
bool TargetList::fetchOrInsert(const Hash &hash, JobPtr &ptr)
{
  LockGuard lock(mutex);
  HMap::iterator it = list.find(hash);
  if(it == list.end())
    {
      assert(ptr);
      assert(!ptr->getInfo()->hasStarted());
      list[hash] = ptr;
      return false;
    }

  ptr = it->second;
  assert(ptr);
  return true;
}
