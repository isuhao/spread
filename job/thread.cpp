#include "thread.hpp"

#include <boost/thread.hpp>
#include <assert.h>

using namespace Jobify;

struct ThreadObj
{
  Job *j;
  void operator()()
  {
    j->run();
    assert(j->getInfo()->isFinished());
    delete j;
  }
};

void Thread::sleep(double seconds)
{
  int msecs = (int)(seconds*1000000);
  boost::this_thread::sleep(boost::posix_time::microseconds(msecs));
}

void Thread::run(Job *j, bool async)
{
  ThreadObj to;
  to.j = j;
  if(async) boost::thread trd(to);
  else to();
}
