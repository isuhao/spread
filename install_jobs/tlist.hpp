#ifndef __SPREAD_TARGET_LIST_HPP
#define __SPREAD_TARGET_LIST_HPP

#include <hash/hash.hpp>
#include <job/job.hpp>
#include "movablelock.hpp"

namespace Spread
{
  struct TargetList
  {
    MovableLock lock() { return MovableLock(new LockGuard(mutex)); }

    void remove(const Hash::DirMap &files);

    // Inserts element if missing. Returns true if found, false if
    // inserted.
    bool fetchOrInsert(const Hash &hash, JobPtr &ptr);

  private:
    typedef std::map<Hash, JobPtr> HMap;
    HMap list;
    Mutex mutex;
  };
}

#endif
