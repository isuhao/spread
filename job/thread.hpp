#ifndef __JOBIFY_THREAD_HPP
#define __JOBIFY_THREAD_HPP

#include "job.hpp"

namespace Jobify
{
  /* This black-box module swallows any Job object, executes it in a
     separate thread, then deletes the object after it returns. (The
     job must have been allocated using 'new'.)

     You can disable threading (run it locally) by setting
     async=false. This may seem pointless compared to just running
     j->run(), but it does provide a consistent interface for both
     cases, and it will also delete the Job before returning.
   */
  struct Thread
  {
    static void run(Job *j, bool async=true);
    static void sleep(double seconds);
  };
}
#endif
