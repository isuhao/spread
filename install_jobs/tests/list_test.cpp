#include "tlist.hpp"

#include <iostream>
#include <job/thread.hpp>

using namespace std;
using namespace Spread;

TargetList list;
bool locked = false;

struct LockJob : Job
{
  void doJob()
  {
    cout << "  LOCKING\n";
    MovableLock lock = list.lock();
    locked = true;
    Thread::sleep(1.5);
    cout << "  UNLOCKING\n";
    setDone();
  }
};

struct NoJob : Job
{
  void doJob()
  {
    cout << "Running nojob... thinking hard ... done!\n";
    setDone();
  }
};

Hash AAA("AAA"), BBB("BBB");

int main()
{
  JobPtr ptr(new NoJob);

  cout << "Insertion:\n";
  cout << list.fetchOrInsert(AAA, ptr) << endl;

  cout << "Fetch (with local lock):\n";
  {
    MovableLock lock = list.lock();
    JobPtr ptr2;
    cout << list.fetchOrInsert(AAA, ptr2) << endl;
    assert(ptr == ptr2);
  }

  cout << "Finishing:\n";
  ptr->run();

  cout << "Removing.\n";
  Hash::DirMap dir;
  dir["abc"] = AAA;
  list.remove(dir);

  cout << "Reinsertion:\n";
  ptr.reset(new LockJob);
  cout << list.fetchOrInsert(AAA, ptr) << endl;
  cout << "Running LockJob.\n";
  Thread::run(ptr);
  while(!locked) Thread::sleep(0.01);
  cout << "Removing entry (will wait for lock)\n";
  list.remove(dir);
  cout << "Entry removed.\n";

  return 0;
}
