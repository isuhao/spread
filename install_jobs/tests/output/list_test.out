Insertion:
0
Fetch (with local lock):
1
Finishing:
Running nojob... thinking hard ... done!
Removing.
Reinsertion:
0
Running LockJob.
  LOCKING
Removing entry (will wait for lock)
  UNLOCKING
Entry removed.
